### PR TITLE
IHE Consistent Presentation of Images

### PR DESCRIPTION
**Pull request by jdnarvaez**
This PR addresses a number of rendering issues and adds some extra functionality. The goal here was to provide functionality to pass the test set for the IHE Consistent Presentation of Images test plan based upon their image display test cases data sets and documents.

MESA/IHE Spec: http://ihewiki.wustl.edu/wiki/index.php/MESA/Image_Display#Image_Display_Test_521:_Consistent_Presentation_of_Images
Test Case Resource: http://ihedoc.wustl.edu/mesasoftware/12.x/display_consistency/index.html

Defects addressed:

- Proper display of images with a pixel representation of 1 (2's complement)
- Perform inversion of images that are MONOCHROME1. This has to be performed in the correct order in the pipeline or else the image will be displayed incorrectly when LUTs are used

Functionality added:

- Auto window level based on output values
- Modality and VOI LUT transformations in web GL as opposed to only supporting slope/intercept and window specifications.
- Support of a `displayedArea` attribute on the viewport that will allow for the specification of the GSPS displayed area module attributes and perform the correct transformations on the image. This includes sub areas, rotations, aspect ratios, etc. The specification of these attributes mirrors the DICOM spec, and I can provide a separate PR that reads a presentation state and generates a cornerstone viewport object, but since most of the DICOM processing is done elsewhere for cornerstone it did not seem appropriate to include here.

This implementation passes *most* test cases. For WebGL, cases 18 & 19 fail for Modality LUTs and cases 5, 9, 10, fail for VOI LUTs still. For the canvas implementation, cases 18 & 19 fail for Modality LUTs. I may very well be overlooking something simple in these cases, but I haven't addressed them yet. 

I have implementations for rendering text annotations, graphic annotations, display shutters, and overlays to facilitate the passage of those test cases in the IHE spec, but these seem suited for the cornerstoneTools repository.

### CODE REVIEW

File: src/internal/generateLut.js [hunk 1]
```diff
@@ -1,11 +1,15 @@
 import getModalityLUT from './getModalityLUT.js';
 import getVOILUT from './getVOILut.js';
+import autoWindow from './autoWindow.js';
 
 /**
  * Creates a LUT used while rendering to convert stored pixel values to
  * display pixels
  *
  * @param {Image} image A Cornerstone Image Object
+ * @param {Viewport} image A Cornerstone Viewport Object
```
Comment (swederik) [inline src/internal/generateLut.js]: Parameter name seems incorrect (should be 'viewport')

**Comment by swederik** [comment]
This is awesome! Thanks!

Presentation state is definitely something that would be awesome to support as well. I'll try to review ASAP, because I already have some questions after skimming it.

**Comment by swederik** [comment]
OK so this PR is pretty huge so I will try to address each of the pieces separately. 

1. **Displayed Area**

The displayedArea viewport property is cool! I don't see any problems with it, so if you split it into a separate PR I would probably just merge that straight away. The only thing I would want is an example demonstrating how (e.g. with some text input boxes for tlhc/trhc/etc) it modifies the display. I have no problem merging it now and adding the example later, though. (Tests would also be nice).

2.  **Auto VOI LUT**

For the auto VOI functionality, there are already a couple related issues (https://github.com/cornerstonejs/cornerstone/pull/144 and https://github.com/cornerstonejs/cornerstoneWADOImageLoader/pull/116). The PR by @adreyfus sets it up so that the WADO loader can populate an array of VOI LUT presets which are then easy to apply when rendering. After reviewing both (sorry @adreyfus this took so long) I am more inclined to merge that one since it is a bit more comprehensive. I think there are some differences between your two PRs (autoWindow: https://github.com/cornerstonejs/cornerstone/pull/211/files#diff-75a498172c64d5d2524c6cd6d7649211R18 vs computeAutoVoi: https://github.com/cornerstonejs/cornerstone/pull/144/files#diff-3e049a6d82baefe3a3ba3a58dee74d17R9) though so there might need to be some more changes.

3. **Signed Pixel Data issues**

The signed pixel data issues were also brought up recently by @kedmon (https://github.com/cornerstonejs/cornerstoneWADOImageLoader/issues/161). This is definitely a problem I want to sort out ASAP and I'm really happy you addressed it in your PR. My main question at this point is whether or not the changes to address images with pixel representation of 1 (getStoredValue / shouldShift / etc) should be done in Cornerstone Core or in the WADO Image Loader. My inclination is that we should push those changes to the WADO Image Loader so that any other procedure using image.getStoredPixels() retrieves pixel values that are already corrected.

That would mean that you would need to push the getStoredValue stuff to the WADO Image Loader and have it run somewhere in the decoding pipeline. I'm not 100% certain where that should go, though. We could do the shifting after image decoding while inside the web worker, somewhere around here (https://github.com/cornerstonejs/cornerstoneWADOImageLoader/blob/master/src/webWorker/decodeTask/decodeTask.js#L84).

I don't know if there is a real benefit to having the getStoredValue / shouldShift stuff in Cornerstone Core, since I guess most image loaders won't require this functionality (feel free to correct me if I'm wrong). It seems more useful to have WADO Image Loader provide correct pixel values automatically. What do you think?

4. **WebGL pipeline changes**

It is a great idea to fix the WebGL pipeline so it can support Modality / VOI LUT transformations. If you perform these transformations (Modality LUT / VOI LUT) in the 'storedPixelDataToImageData' functions, though, I guess you need to disable them in the actual shaders themselves (e.g. https://github.com/cornerstonejs/cornerstone/blob/master/src/webgl/shaders/int16.js#L59).

One thing we should work towards would be automated testing for the IHE Consistent Presentation of Images. I guess we could do some stored pixel value comparisons in the WADO image loader and then actual display consistency verification in Core. Comparing the test output to a known RGBA array shouldn't be too hard, and it looks like a lot of the tests are meant to display "a normal SMPTE pattern".

**Comment by jdnarvaez** [comment]
And now I remember why I put the stored value stuff in core: after performing the two's complement conversion the minMax of the image would change, resulting in a different value for accurate windowing. I can still but the initial transformation in the WADO image loader and expect that the eventual auto windowing function calculates the min/max correctly.

**Comment by swederik** [comment]
Sounds good! 

Hmm we calculate min / max in WADO Image Loader as well, though: https://github.com/cornerstonejs/cornerstoneWADOImageLoader/blob/master/src/webWorker/decodeTask/decodeTask.js#L88 so if you do the two's complement conversion before that point, I guess it's okay.

I'll look into setting up tests for the IHE dataset.

**Comment by jdnarvaez** [comment]
https://github.com/cornerstonejs/cornerstone/pull/211

**Comment by jdnarvaez** [comment]
https://github.com/cornerstonejs/cornerstoneWADOImageLoader/pull/168

**Comment by Zaid-Safadi** [comment]
I'd like to get the features of this PR merged in the master gradually starting with the "**displayedArea**" as it is the simplest and most straightforward. Here is the plan:
1. Create a new branch "PR-211" and merge all this code into.
2. Create a new branch "displayedArea"
3. cherry-pick the two commits 96ba959b7f38f6e5e244bbe62ddea04d3ed4400f and 3b8901ae781dadedd1f55831df2cbe3288ad49b6 and merge them into the "displayedArea" branch
4. Do any necessary cleanup then merge to "master"
5. Loop through the same process working off "PR-211" branch until all features has been merged

@jdnarvaez it will be very helpful if you can provide some test cases for the "displayedArea" feature to make sure all merged code is working properly.

**Comment by jdnarvaez** [comment]
@Zaid-Safadi 

I would be more than happy to move things out piece by piece, I'm just a little limited on time currently.

**Comment by Zaid-Safadi** [comment]
@jdnarvaez thank you for the offer, this will be fantastic!

I will still merge the displayedArea feature as this seems a quick one and maybe let you test and review before merging to master.

For the other features and specifically the "Auto VOI LUT" I agree that your code is handling other cases that are not addressed in #144. I will do some clean-up to this in the voi-lut branch and maybe you can update that branch then with your changes.

The rest then can be handled easily :)

**Comment by Zaid-Safadi** [comment]
The new "displayed-area" branch is now created with the "Displayed Area" feature changes cherry-picked into it. @jdnarvaez , please feel free to validate the branch and provide your comments. I still like to perform some tests before merging to master.

**Comment by Zaid-Safadi** [comment]
I have updated the HTML example "**viewport**" in the branch [**displayed-area**](https://github.com/cornerstonejs/cornerstone/commit/b4fd578af5e0644be0993735956195fd7d03be0a) to show and test this feature.

@jdnarvaez, I need you to verify that this example is working as you expect the code to work. I have done quick test and yes the viewport is changing but I am not sure if this is the intended behavior 

We should also document that this is only for "SCALE TO FIT", do you have any plans to add support for "TRUE SIZE" and "MAGNIFY"?

display-area

**Pull request merged**